United Hatzala data integration

omerxx opened the pull request:
#427: Integrating United Hatzala data.
The DB population and csv parsing are both done in the new `united.py`
I created a different marker for their accidents as they have much 'thinner' data.
Accidents are only given in severities `קשה` and `קלה` so there are no fatal markers.

* Update: Accidents are being imported using `importmail.py` which downloads the .csv files directly from gmail. the script can be called directly or automatically in `united.py`. If it finds `static/data/united/` empty it downloads all accidents history. If it's not, then only the last file is downloaded.

@danielhers - Can this PR be deployed to a staging app with the latest accidents file?

Example shot:
![image](https://cloud.githubusercontent.com/assets/12237955/9875541/773dfa60-5bb7-11e5-87ba-286bf54b74f7.png)


<!---
@huboard:{"order":438.0,"milestone_order":435,"custom_state":""}
-->

danielhers commented on models.py:
```diff
@@ -202,6 +201,12 @@ def serialize(self, is_thin=False):
                 "secondaryStreet": self.secondaryStreet,
                 "junction": self.junction,
             })
+
```
Requires comment, plus named constant for `2`

omerxx replied:
Done

danielhers commented on static/js/constants.js:
```diff
@@ -103,6 +105,13 @@ function getIcon(accidentType, severity) {
     try {
         if (accidentType == "multiple") {
             icon = MULTIPLE_ICONS[severity];
+        } else if (accidentType == 21) {
```
ditto

omerxx replied:
Done

danielhers commented on static/js/marker.js:
```diff
@@ -37,6 +38,45 @@ var MarkerView = Backbone.View.extend({
 
+        if (this.model.get("provider_code") == 2) {     // United Hatzala accidents marker structure
+            var united_icon;
+            if (this.model.get("severity") == 3) {
```
Use the function from `static/js/constants.js`

omerxx replied:
Done

danielhers commented on static/js/marker.js:
```diff
@@ -37,6 +38,45 @@ var MarkerView = Backbone.View.extend({
+            if (isRetina) {
+                this.marker.setIcon( app.retinaIconsResize(addRetinaSuffix(united_icon)) );
+            } else {
+                this.marker.setIcon( UNITED_HATZALA_ICON );
```
What about considering severity in this case?

omerxx replied:
Right :)
Fixed by using `constants.js` function

danielhers commented on static/js/marker.js:
```diff
@@ -37,6 +38,45 @@ var MarkerView = Backbone.View.extend({
+            app.oms.addMarker(this.marker);
+            this.$el.html($("#united-marker-content-template").html());
+
+            this.$el.find(".title").text(this.marker.get("title"));
```
Aren't many of these done also when `provider_code` is not `2`? Why repeat them?

omerxx replied:
`setMap` and `view` are the actual ones the repeat in all 3 cases. I took them out.

danielhers commented on united.py:
```diff
@@ -0,0 +1,103 @@
+    :return: Python datetime object
+    """
+    time = datetime.strptime(created[:-3], '%m/%d/%Y %I:%M:%S')
+    year = int(time.strftime('20%y'))
```
Why not `time.year` etc.?

omerxx replied:
Wouldn't that simply return the current year..?

danielhers replied:
`time` here is used as a variable, you're not importing it

omerxx replied:
Right... confused with datetime. Fixed.

danielhers commented on united.py:
```diff
@@ -0,0 +1,103 @@
+    db = SQLAlchemy(app)
+    try:
+        accidents = list(create_accidents(path))
+        db.session.execute(Marker.__table__.insert().prefix_with("OR IGNORE"), accidents)
```
I think `OR IGNORE` doesn't work on PostgreSQL

omerxx replied:
I couldn't understand the SOF explanation. Would need some help on this..

omerxx replied:
Considering the use of python's `poplib` and download MIME types (attachments) to the server.

danielhers commented on static/js/marker.js:
```diff
@@ -21,22 +21,51 @@ var MarkerView = Backbone.View.extend({
             return this;
         }
 
+        if (this.model.get("provider_code") == 2) {     // United Hatzala accidents marker structure
```
Use the new constant

omerxx commented on united.py:
```diff
@@ -0,0 +1,103 @@
+                continue
+            if first_line:
+                first_line = False
+                if accident[0] == "":
```
note to self --> use csvmap

omerxx commented on united.py:
```diff
@@ -0,0 +1,103 @@
+                if accident[0] == "":
+                    print "\t\tEmpty File!"
+                    continue
+            if accident[1] == "" or accident[2] == "":
```
csvmap

omerxx commented:
Requirements (will be fixed soon):
* Add apscheduler integration to `main.py`
* Fix imported attachments count using `total`
* Implement command line arguments to allow `united.py` run without downloading from gmail. Also implement an arg to force `importmail.py` download all mail even if its folder is not empty. (Only incremental files will be actually downloaded)

omerxx commented:
Added APScheduler (2.1.2) - scheduled incremental download every 12 hours
* Since data is 3 hours old to begin with, and many files during the day are empty I came to conclude that the best practice would be a 12 hour scheduled incremental download.

Added optional command line arguments (having only last file downloaded, password via cmd line)
Changed import to always be incremental (only importing last hour file is not a default)
Fixed automation issues

@danielhers - I think what's left is just the postgresql UPSERT and we can let it out

danielhers commented:
Working on the UPSERT

danielhers commented on importmail.py:
```diff
@@ -0,0 +1,110 @@
+def main():
+    maildir = 'united-hatzala/data'
+    detach_dir = 'static/data/united'
+    username = 'anyway@anyway.co.il'
```
I think this should also be an environment variable.

danielhers commented on models.py:
```diff
@@ -70,6 +70,7 @@ def __unicode__(self):
 
 MARKER_TYPE_ACCIDENT = 1
 MARKER_TYPE_DISCUSSION = 2
+UNITED_HATZALA_CODE = 2
```
I wonder why we need both this and `PROVIDER_CODE_UNITED_HATZALA` from `constants.js`.

omerxx replied:
Both of them were set to your suggestion... 
Would you like to remove any of them?

danielhers replied:
I'm not always right :)
Why would there be two constants?

omerxx replied:
Because one serves the python side while the other is a js constant..
Both of them are improving code readability...

danielhers replied:
Right, of course. Thought both are js for some reason...

omerxx replied:
:)

danielhers commented on importmail.py:
```diff
@@ -0,0 +1,110 @@
+                    break
+                total += 1
+                print 'Currently loading: ' + filename
+                sys.stdout.write("\033[F")
```
If a shorter filename is printed, the previous one's last character(s) are still shown. Maybe write as many space character as the previous filename's length before writing this filename.

omerxx commented:
Fixed rejections:

* User name is now an environment variable MAILUSER (or can be set via --username)
* Fixed string overwriting while downloading files

danielhers commented on united.py:
```diff
@@ -0,0 +1,114 @@
+    Calls importmail.py prior to importing to DB
+    """
+    parser = argparse.ArgumentParser()
+    parser.add_argument('--light', action='store_true', default=False)
```
add description

danielhers commented on importmail.py:
```diff
@@ -0,0 +1,121 @@
+        imapsession.select(maildir)
+        typ, data = imapsession.search(None, 'ALL')
+    except:
+        print 'Error searching given maibox: %s' % maildir
```
typo: maibox

danielhers commented on static/js/constants.js:
```diff
@@ -103,6 +108,9 @@ function getIcon(accidentType, severity) {
             icon = MULTIPLE_ICONS[severity];
+
+        } else if (accidentType == ACCIDENT_TYPE_UNITED_HATZALA) {
+            icon = ICONS[severity][ACCIDENT_TYPE_UNITED_HATZALA];
```
Why not just add an entry to `ACCIDENT_MINOR_TYPE_TO_TYPE` rather than adding an `if`?

danielhers commented on importmail.py:
```diff
@@ -0,0 +1,121 @@
+        else:
+            username = os.environ['MAILUSER']     # Set and environment variable MAILUSER with the password
+    except:
+        print "Please set env var MAILUSER, or provide one using importmail.py --username <username>"
```
The second half of this doesn't make sense if this was called from `united.py`. I think in that case it should suggest the user to add `--light` if they want to load from their local directory.

danielhers replied:
But actually, the command line arguments work if you ran `united.py` and not `importmail.py` directly.
You should use `argv[0]` rather than "importmail.py" in the message.

omerxx replied:
Actually I don't want to cause users mistake and run `united.py --username ...`
I cam simply delete the second part.. 
The arguments are mentions in the documentation if the user should have questions

danielhers replied:
But why would that be a mistake?

omerxx replied:
Because `username` and `password` are `importmail.py` arguments..

danielhers commented on static/js/constants.js:
```diff
@@ -103,7 +109,7 @@ function getIcon(accidentType, severity) {
         if (accidentType == "multiple") {
             icon = MULTIPLE_ICONS[severity];
-        } else {
+        }  else {
```
space :x

danielhers commented on united.py:
```diff
@@ -0,0 +1,114 @@
+    """
+    Calls importmail.py prior to importing to DB
+    """
+    parser = argparse.ArgumentParser(description='Allow importing without downloading any new files')
```
should be on the arg, not the parser (next line)

omerxx replied:
the arg doesn't accept `description`

danielhers replied:
Sorry, it's called `help`

danielhers commented on united.py:
```diff
@@ -0,0 +1,114 @@
+                "severity": 2 if u"קשה" in unicode(accident[csvmap["type"]], encoding='utf-8') else 3,
+                "locationAccuracy": 1,
+                "subtype": 21,           # New subtype for United Hatzala
+                "type": unicode(accident[csvmap["casualties"]], encoding='utf-8'),
```
Isn't `type` an integer field? This fails because `casualities` is empty in one of the entries and so it's not an integer.
I thought `type` should always be `MARKER_TYPE_ACCIDENT`...

omerxx replied:
It should.
But since this is a slightly different kind of data I had to 'stuck' the casualties somewhere.
I'll address it as integer

danielhers replied:
If it really has no sensible column to put it in, we can add a new column with a default value of `-1` ("unknown"), for example.

omerxx replied:
It would be a huge waste in my opinion. Only few of the entries have casualties..
If I'll use type as integer wouldn't it solve postgresql issues with empty data?

danielhers replied:
If you'll make sure it's not empty, I suppose. It's a bit ugly though... The `type` column distinguishes accident markers from discussion markers. If its value is 2 (discussion) then we'll show this as a discussion marker which is wrong. Also, I think the default filtering in `/markers` is only taking accident markers (1). Perhaps there's another column that can be used.

omerxx replied:
It's incredibly meaningless.. 25 accidents of ~2000 united hatzala entries actually have a value for that.
I could always use `mainStreet` / `secondaryStreet` / `junction`. Those are `TEXT` and can store the strings since they are not in use with united hatzala.
What do you say?

danielhers replied:
If it's so rare then perhaps it's not even worth saving at all...
Note that there's also a `description` field (that used to contain all the information we weren't loading to its own column). How about it? Or something like the `intactness` column which is also an integer but not critical like `type`.

omerxx replied:
As far as I remember `description` is being used by the IOS and Android apps..
But I agree it's not worth saving. We can take it out for now

danielhers commented on united.py:
```diff
@@ -0,0 +1,114 @@
+                "longitude": accident[csvmap["long"]],
+                "created": parse_date(accident[csvmap["time"]]),
+                "provider_code": 2,
+                "id": int(''.join(x for x in accident[csvmap["time"]] if x.isdigit()) + str(inc)),
```
This gives `integer out of range` on PostgreSQL. Should shorten it so it is less than `2147483647`.

danielhers replied:
I'll ask United if they can add their pk to the csv

omerxx replied:
I'll shorten it. That's no problem

omerxx commented:
I think It should run now on postgreSQL (after your upsert fix is merged).

danielhers commented:
I'll try merging then and see.
By the way, it's not really upsert... It's just like OR IGNORE - only inserting rows that haven't been inserted yet.

omerxx commented:
Yeah I can see now.
So it should be applicable both for sqlite and posgresql?

danielhers commented:
Yes, there's nothing DB-specific there.

danielhers commented:
Merged.
